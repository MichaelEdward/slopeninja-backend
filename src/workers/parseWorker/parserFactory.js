import cheerio from 'cheerio';

export const removeBackSlashes = (text) => {
  const decoded = text ? text.toString().replace(/\\/g, '') : null;
  return decoded;
}

export const createHtmlParser = (key, parser, preParser = t => t) => async (html) => {
  const $ = cheerio.load(preParser(html));

  const keyData = await parser($);

  return {
    [key]: keyData
  };
}

export const createJSONParser = (key, parser) => async (data = "{}") => {
  let json = JSON.parse(data);

  let d;
  if (json) {
    d = json;
  }

  const keyData = await parser(d);

  return {
    [key]: keyData
  };
}
