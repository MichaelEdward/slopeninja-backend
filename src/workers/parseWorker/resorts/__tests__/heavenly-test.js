import fs from 'fs';
import {
  parseHeavenlySnow,
  parseHeavenlyLifts,
  parseHeavenlyLiftList,
  parseHeavenlyTrailList,
  parseHeavenlyTrails
} from '../heavenly';
import { createHtmlParser, removeBackSlashes } from '../../parserFactory';

test('fetches Heavenly snow data correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/fixtures/heavenly-weather.html.input`);

  const resortData = await createHtmlParser('snow', parseHeavenlySnow)(htmlText);
  expect(resortData).toEqual({
    snow: {
      status: 'closed',
      weatherIcon: null,
      temprature: null,
      baseCondition: null,
      newSnow: 0,
      snowDepthBase: null,
      snowDepthSummit: 0,
    },
  });
})

test('fetches all null for nonexisting snow values', async () => {
  const resortData = await createHtmlParser('snow', parseHeavenlySnow)('<html></html>');
  expect(resortData).toEqual({
    snow: {
      status: null,
      weatherIcon: null,
      temprature: null,
      baseCondition: null,
      newSnow: null,
      snowDepthBase: null,
      snowDepthSummit: null,
    }
  });
});

test('fetches Heavenly lifts data correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/fixtures/heavenly-weather.html.input`);
  const resortData = await createHtmlParser('lifts', parseHeavenlyLifts)(htmlText);
  expect(resortData).toEqual({
    lifts: {
      total: null,
      open: null,
    }
  });
});

test('fetches all null for nonexisting lift values', async () => {
  const resortData = await createHtmlParser('lifts', parseHeavenlyLifts)('<html></html>');
  expect(resortData).toEqual({
    lifts: {
      total: null,
      open: null,
    }
  });
});

test('fetches Heavenly trails data correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/fixtures/heavenly-weather.html.input`);
  const resortData = await createHtmlParser('trails', parseHeavenlyTrails)(htmlText);
  expect(resortData).toEqual({
    trails: {
      total: null,
      open: null,
    }
  });
});

test('fetches all null for nonexisting trails values', async () => {
  const resortData = await createHtmlParser('trails', parseHeavenlyTrails)('<html></html>');
  expect(resortData).toEqual({
    trails: {
      total: null,
      open: null,
    }
  });
});

test('fetches Heavenly lift list correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/fixtures/heavenly-lifts.html.input`);
  const resortData = await createHtmlParser(
    'liftList',
    parseHeavenlyLiftList,
    removeBackSlashes,
  )(htmlText);
  expect(resortData).toMatchSnapshot();
});

test('fetches all null for nonexisting lift list values', async () => {
  const resortData = await createHtmlParser(
    'liftList',
    parseHeavenlyLiftList,
    removeBackSlashes,
  )('<html></html>');
  expect(resortData).toMatchObject({ liftList: [] });
});

test('fetches Heavenly trail list correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/fixtures/heavenly-lifts.html.input`);
  const resortData = await createHtmlParser(
    'trailList',
    parseHeavenlyTrailList,
    removeBackSlashes,
  )(htmlText);
  expect(resortData).toMatchSnapshot();
});

test('fetches all null for nonexisting lift list values', async () => {
  const resortData = await createHtmlParser(
    'trailList',
    parseHeavenlyTrailList,
    removeBackSlashes,
  )('<html></html>');
  expect(resortData).toMatchObject({ trailList: [] });
});
