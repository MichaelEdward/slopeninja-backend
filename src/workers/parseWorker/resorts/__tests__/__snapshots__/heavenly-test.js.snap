// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches Heavenly lift list correctly 1`] = `
Object {
  "lifts": Array [
    Object {
      "category": null,
      "name": "Tamarack",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Big Easy",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Dipper",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Comet",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Mott",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Galaxy",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Olympic",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "North Bowl",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Boulder",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Stagecoach",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Gondola",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Sky",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Canyon",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Powderbowl",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Groove",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Patsy's",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "World Cup",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "First Ride",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Gunbarrel",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Aerial Tram",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "DMZ Carpet",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Enchanted Carpet",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Mitey Mite 1",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Mitey Mite 2",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Pioneer Mitey-Mite",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Boulder Carpet",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Red Fir Mitey Mite",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Bear Cave Carpet",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Tubing Lift",
      "status": "closed",
    },
  ],
}
`;

exports[`fetches Heavenly trail list correctly 1`] = `
Object {
  "trails": Array [
    Object {
      "category": "Top Of Gondola",
      "level": "beginner",
      "name": "Big Easy",
      "status": "closed",
    },
    Object {
      "category": "Top Of Gondola",
      "level": "beginner",
      "name": "Black Bear Hollow",
      "status": "closed",
    },
    Object {
      "category": "Top Of Gondola",
      "level": "beginner",
      "name": "Easy Street",
      "status": "closed",
    },
    Object {
      "category": "Top Of Gondola",
      "level": "beginner",
      "name": "Tamarack Return",
      "status": "closed",
    },
    Object {
      "category": "Top Of Gondola",
      "level": "intermediate",
      "name": "49er",
      "status": "closed",
    },
    Object {
      "category": "Top Of Gondola",
      "level": "intermediate",
      "name": "California Trail",
      "status": "closed",
    },
    Object {
      "category": "Top Of Gondola",
      "level": "intermediate",
      "name": "California Trail -Lower",
      "status": "closed",
    },
    Object {
      "category": "Top Of Gondola",
      "level": "intermediate",
      "name": "Samu2019s Dream",
      "status": "closed",
    },
    Object {
      "category": "Top Of Gondola",
      "level": "intermediate",
      "name": "Silver Spur",
      "status": "closed",
    },
    Object {
      "category": "Top Of Gondola",
      "level": "intermediate",
      "name": "VonSchmidt's",
      "status": "closed",
    },
    Object {
      "category": "Top Of Gondola",
      "level": "advanced",
      "name": "Aries Woods",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "beginner",
      "name": "First Ride Run",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "beginner",
      "name": "Groove Terrain Park",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "beginner",
      "name": "Maggie's",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "beginner",
      "name": "Mombo",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "beginner",
      "name": "Mombo Meadows",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "beginner",
      "name": "Patsy's",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "beginner",
      "name": "Swing Trail",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "intermediate",
      "name": "Advanced Round-a-bout",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "intermediate",
      "name": "Cat Track",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "intermediate",
      "name": "EpicMix Racing",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "intermediate",
      "name": "High Five",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "intermediate",
      "name": "High Roller Left",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "intermediate",
      "name": "High Roller Right",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "intermediate",
      "name": "Liz's-Lower",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "intermediate",
      "name": "Liz's-Upper",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "intermediate",
      "name": "Powderbowl Run",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "intermediate",
      "name": "Powderbowl Woods",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "intermediate",
      "name": "Ridge Run  ",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "intermediate",
      "name": "Ridgeway",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "intermediate",
      "name": "Round-a-bout",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "intermediate",
      "name": "Round-a-bout Swing",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "intermediate",
      "name": "Sky Chute",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "intermediate",
      "name": "Skyline Trail",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "intermediate",
      "name": "Upper Mombo",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "intermediate",
      "name": "World Cup",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "advanced",
      "name": "Double Down Lower",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "advanced",
      "name": "Double Down Upper",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "advanced",
      "name": "Ellie's Swing",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "advanced",
      "name": "Ellie's-Lower ",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "advanced",
      "name": "Ellie's-Upper ",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "advanced",
      "name": "Express Line",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "advanced",
      "name": "Fall Line",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "advanced",
      "name": "Pinnacles Left",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "advanced",
      "name": "Pinnacles Right",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "advanced",
      "name": "Pistol",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "advanced",
      "name": "Powder Line",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "advanced",
      "name": "Ridge Bowl",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "advanced",
      "name": "Ski Way glades",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "advanced",
      "name": "Sky Canyon",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "advanced",
      "name": "Waterfall",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "advanced",
      "name": "West Bowl",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "expert",
      "name": "East Bowl",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "expert",
      "name": "Gunbarrel-Lower ",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "expert",
      "name": "Gunbarrel-Upper",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": "expert",
      "name": "The Face",
      "status": "closed",
    },
    Object {
      "category": "California",
      "level": null,
      "name": "Groove Park",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "beginner",
      "name": "Boulder Bowl",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "beginner",
      "name": "Easiest Way",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "$100 Saddle",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Aries",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Big Dipper-Lower",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Big Dipper-Upper",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Bonanza",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Boulder Chute",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "California Tr.-Up",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "California Trail u2013 Lower",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Cloud Nine",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Comet",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Comstock Lower",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Comstock Upper",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Cosmic Wave",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Crossover",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Emily's Run",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Galaxy",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Galaxy Line",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Jack's",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Lower Dipper Rtn. ",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Meteor",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Mineshaft Lower",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Mineshaft Upper",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Nevada Trail Lower",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Nevada Trail Upper",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Nova",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Olympic DH-Lower ",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Olympic DH-Mid",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Olympic DH-Upper ",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Orion's",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Orion's Belt",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Outlaw ",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Pepi's",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Perimeter",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Ponderosa",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Sand Dunes",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Stagecoach Return",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Stagecoach Upper",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Stagecoach-Lower ",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Stagecoach-Middle ",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "The Pines",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Upper Dipper Return",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Upper North Bowl",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "intermediate",
      "name": "Upper Perimeter",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "advanced",
      "name": "Bohemian Grove",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "advanced",
      "name": "Dipper Bowl",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "advanced",
      "name": "Dipper Line",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "advanced",
      "name": "Dipper Woods",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "advanced",
      "name": "Little Dipper",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "advanced",
      "name": "Milky Way",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "advanced",
      "name": "Milky Way Bowl",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "advanced",
      "name": "North  Bowl",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "advanced",
      "name": "Stagecoach woods",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "advanced",
      "name": "Stageline",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "expert",
      "name": "Killebrew Canyon",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": "expert",
      "name": "Mott Canyon",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": null,
      "name": "Ante Up Terrain Park",
      "status": "closed",
    },
    Object {
      "category": "Nevada",
      "level": null,
      "name": "High Roller Terrain Park",
      "status": "closed",
    },
  ],
}
`;
